### PR TITLE
Gracefully handle exceptions in code fixer creation

### LINKED ISSUE #45851: Exception in CodeFixService for abstract type with ExportCodeFixProviderAttribute
**Version Used**: Latest 16.7 int preview build

This only seems to occur for Workspace/VSIX based code fix providers. We handle the exception gracefully from NuGet based code fix provider. We should be more resilient here.
```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: System.MemberAccessException: Cannot create an instance of Microsoft.NetCore.Analyzers.Performance.PreferIsEmptyOverCountFixer because it is an abstract class.
   at System.Reflection.RuntimeConstructorInfo.CheckCanCreateInstance(Type declaringType, Boolean isVarArg)
   at System.Reflection.RuntimeConstructorInfo.ThrowNoInvokeException()
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.VisualStudio.Composition.ReflectionHelpers.Instantiate(MethodBase ctorOrFactoryMethod, Object[] arguments)
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.RuntimePartLifecycleTracker.CreateValue()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.Create()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveNext(PartLifecycleState nextState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveToState(PartLifecycleState requiredState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.GetValueReadyToExpose()
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.<>c__DisplayClass15_0.<GetExportedValueHelper>b__0()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.<>c__DisplayClass15_0.<GetExportedValueHelper>b__0()
   at Microsoft.VisualStudio.Composition.DelegateServices.<>c__DisplayClass2_0`1.<As>b__0()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Lazy`1.get_Value()
   at Microsoft.CodeAnalysis.CodeFixes.CodeFixService.<>c__DisplayClass31_1.<GetFixerPerLanguageMap>b__0()
   at System.Lazy`1.CreateValue()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Lazy`1.get_Value()
   at Microsoft.CodeAnalysis.CodeFixes.CodeFixService.<AppendFixesAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.CodeAnalysis.CodeFixes.CodeFixService.<GetFixesAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Roslyn.Utilities.TaskExtensions.WaitAndGetResult_CanCallOnBackground[T](Task`1 task, CancellationToken cancellationToken)
```

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #45851

This happened because an abstract fixer type had ExportCodeFixProvider attribute, which caused VS Crash on Ctrl + Dot operation. See https://github.com/dotnet/roslyn-analyzers/pull/3843

### CODE REVIEW

File: src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs [hunk 1]
```diff
@@ -762,17 +762,28 @@ private static ImmutableArray<string> GetAndTestFixableDiagnosticIds(CodeFixProv
...
-                        foreach (var id in this.GetFixableDiagnosticIds(fixer.Value, extensionManager))
+                        CodeFixProvider fixer;
+                        try
+                        {
+                            fixer = lazyFixer.Value;
+                        }
+                        catch
+                        {
+                            // Gracefully handle exceptions in creating fixer instance.
+                            continue;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:775]: can we log/diagnostic (i.e. an AD diagnostic) these or yellow-bar somehow?  Having stuff fail to create and trying to track it down is a nightmare at times.

Comment (mavasani) [reply src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:775]: Sure, I think we can show an infobar with type name and exception stack trace. Let me add that part.

Comment (mavasani) [reply src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:831]: Done. Now we show the below:

image

image

File: src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs [hunk 2]
```diff
@@ -42,13 +42,15 @@ internal partial class CodeFixService : ForegroundThreadAffinitizedObject, ICode
 
         private readonly IDiagnosticAnalyzerService _diagnosticService;
 
-        private readonly ImmutableDictionary<LanguageKind, Lazy<ImmutableDictionary<DiagnosticId, ImmutableArray<CodeFixProvider>>>> _workspaceFixersMap;
+        private readonly Func<Workspace, ImmutableDictionary<LanguageKind, Lazy<ImmutableDictionary<DiagnosticId, ImmutableArray<CodeFixProvider>>>>> _getWorkspaceFixersMap;
```
Comment (jmarolf) [inline src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:45]: wew thats very generic

File: src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs [hunk 3]
```diff
@@ -266,6 +268,63 @@ public async Task<Document> ApplyCodeFixesForSpecificDiagnosticIdAsync(Document
...
+                fixer = lazyFixer.Value;
+                return true;
+            }
+            catch (Exception ex)
+            {
+                // Gracefully handle exceptions in creating fixer instance.
+                // Show info bar for exception, if needed.
+                if (showInfoBarOnException)
+                {
+                    var errorReportingService = workspace.Services.GetRequiredService<IErrorReportingService>();
```
Comment (jmarolf) [inline src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:310]: doesn't look like the `VisualStudioErrorReportingService` which this is going to call in VS' case will log failure telemetry. Is that something we want to do for this case?

Comment (mavasani) [reply src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:310]: Seems like that should be an independent change made to `VisualStudioErrorReportingService`, as we probably want telemetry from all clients invoking this?

Comment (mavasani) [reply src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:310]: Actually, it seems the CodeFixService already MEF imports the error loggers, so I will update the code to log exception here. Thanks!

**Comment by mavasani** [comment]
@jmarolf @CyrusNajmabadi any further feedback? Thanks.

File: src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs [hunk 4]
```diff
@@ -266,6 +268,68 @@ public async Task<Document> ApplyCodeFixesForSpecificDiagnosticIdAsync(Document
...
+                        : EditorFeaturesResources.Error_creating_instance_of_CodeFixProvider;
+
+                    errorReportingService.ShowErrorInfoInActiveView(
+                        message,
+                        new InfoBarUI(
+                            WorkspacesResources.Show_Stack_Trace,
+                            InfoBarUI.UIKind.HyperLink,
+                            () => errorReportingService.ShowDetailedErrorInfo(ex), closeAfterAction: true));
+
+                    foreach (var errorLogger in _errorLoggers)
```
Comment (jmarolf) [inline src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:322]: shouldn't we log this even if `showInfoBarOnException` is set to `false`?

Comment (mavasani) [reply src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:322]: Let me rename the variable - we don't want to log exception or show info bar when invoked from GetFixableDiagnosticIDs context.

**Pull request merged**